status/reporters.go

```diff
@@ -413,3 +487,112 @@ func (r *GitHubReporter) ReportStatus(k8sClient client.Client, ctx context.Conte
 
 	return nil
 }
+
+// ReportStatusForSnapshot creates CheckRuns when using GitHub App integration.
+// When using GitHub webhook integration a commit status
+func (r *GitHubReporter) ReportStatusForSnapshot(k8sClient client.Client, ctx context.Context, snapshot *applicationapiv1alpha1.Snapshot, details *[]gitops.IntegrationTestStatusDetail) error {
+	if !metadata.HasLabelWithValue(snapshot, gitops.PipelineAsCodeEventTypeLabel, gitops.PipelineAsCodePullRequestType) {
+		return nil
+	}
+
+	// Existence of the Pipelines as Code installation ID annotation signals configuration using GitHub App integration.
+	// If it doesn't exist, GitHub webhook integration is configured.
+	if metadata.HasAnnotation(snapshot, gitops.PipelineAsCodeInstallationIDAnnotation) {
+		creds, err := r.getAppCredentials(ctx, snapshot)
+		if err != nil {
+			return err
+		}
+
+		token, err := r.client.CreateAppInstallationToken(ctx, creds.AppID, creds.InstallationID, creds.PrivateKey)
+		if err != nil {
+			return err
+		}
+
+		r.client.SetOAuthToken(ctx, token)
+
+		labels := snapshot.GetLabels()
+
+		owner, found := labels[gitops.PipelineAsCodeURLOrgLabel]
+		if !found {
+			return fmt.Errorf("snapshot label not found %q", gitops.PipelineAsCodeURLOrgLabel)
+		}
+
+		repo, found := labels[gitops.PipelineAsCodeURLRepositoryLabel]
+		if !found {
+			return fmt.Errorf("snapshot label not found %q", gitops.PipelineAsCodeURLRepositoryLabel)
+		}
+
+		sha, found := labels[gitops.PipelineAsCodeSHALabel]
+		if !found {
+			return fmt.Errorf("snapshot label not found %q", gitops.PipelineAsCodeSHALabel)
+		}
+
+		for _, integrationTestStatusDetail := range *details {
+			// We report the final status of handling a integrationTestScenario only
+			// such as TestFail, TestPassed, DeploymentError, EnvironmentProvisionError
+			integrationTestStatusDetail := integrationTestStatusDetail // G601
+			if integrationTestStatusDetail.Status == gitops.IntegrationTestStatusEnvironmentProvisionError || integrationTestStatusDetail.Status == gitops.IntegrationTestStatusDeploymentError || integrationTestStatusDetail.Status == gitops.IntegrationTestStatusTestFail || integrationTestStatusDetail.Status == gitops.IntegrationTestStatusTestPassed {
+				checkRun, err := r.createCheckRunAdapterForSnapshot(k8sClient, ctx, snapshot, integrationTestStatusDetail, owner, repo, sha)
+				if err != nil {
+					return err
+				}
+
+				checkRunID, updateIsNeeded, err := r.client.GetCheckRunIDAndIfUpdateNeeded(ctx, owner, repo, sha, checkRun.ExternalID, checkRun.CompletionTime, creds.AppID)
```

Comment:
I would revise the functions to load Checkrun list and CommitStatus list once for each of them then to compare with the one to be created/updated to reduce the times of communicating with Github.

Reply:
Done.